### DISCUSSION TITLE
Code execution for other languages (i.e. other than R, Python, and Julia)

### DISCUSSION THREAD

**Discussion by cmhh** [quarto-dev/quarto-cli#847] [Q&A]
Since Python and Julia code can be executed via Jupyter kernels, I got curious and tried some others.  To my suprise (I don't think I saw this documented anywhere), the following worked just fine for me:

~~~md
---
...
jupyter: scala213
---

```{scala}
println("Hello, World!")
``` 

~~~

Is this something that is expected / intended?  If I make use of this functionality, do we expect it to break at some point?

**Comment by jjallaire** [comment]
Yes, this is indeed expected and intended! We don't play up the other kernels because we haven't tested them but they should definitely all work.

**Reply by cmhh** [reply]
I mean, this is amazing.  And it means I can mix languages in the same document.  I'm going to try IHaskell as soon as I finish work!

**Comment by jjallaire** [comment]
Note that you can only have a single kernel per document. There are some ways around this e.g. Python cell magics but in general the design of Jupyter is one language per notebook.

**Reply by cmhh** [reply]
Okay, that makes sense.  I thought it was working, but I think the caches from multiple runs were getting mashed together and it only looked that way.  I guess it's relatively rare that you'd need to use more than1 language, and it's still nice to be able to use the almond kernel in particular.

Still, I _think_ I've tested it correctly, and it at least appears you can used a different kernel for each chapter of a book (and so I assume a blog also)?  That's a good compromise--especially in the case of a blog.

**Reply by cderv** [reply]
@cmhh depending on your use case, if using R can be an option, the computation engine used with Quarto could be **knitr** which will allow you to mix code chunks in different language, as **knitr** supports running different chunk language engine in the same document. R, Python, Julia and others with native support or through R packages. 

It is not quite the same but in some case that could help have different language in the same doc. But you may already know that. 😄

**Reply by cmhh** [reply]
Yes.  I was writing a user guide for analysts at work and wanted to include simple database access examples in a few different languages.  I was _hoping_ to use R to handle R, Python, and Julia via reticulate and JuliaCall, and then Scala via the almond kernel.  Caching aside, I can do R, Python, and Julia; or Scala... but not all 4.  Some might say mixing 4 runtimes in a single document is excessive, of course. :)

**Reply by cderv** [reply]
With **knitr** you should be able to do `scala` too, it is recognized as an interpreted engine (https://bookdown.org/yihui/rmarkdown-cookbook/other-languages.html). This means it will be run as a command line call passing the chunk content with `-e` options. 

But it is quite limited as you only get the printed result in the chunk output. So you need to export result from scala in a way loadable in R to process further. You can really use object in R as with other integration; 

However, there seems to be a package that offer scala support in R that could be used with **knitr** : https://www.zstat.pl/2018/07/27/scala-in-knitr/
though I never tried it. 

Just sharing this in case it could help

**Reply by cmhh** [reply]
Thanks for the suggestion.  I have tried rscala in the past, but it doesn't seem to have had any work done since 2020, so wondered if I should think of it as deprecated.  I'll have a little fiddle, though.  I don't need interop between R and Scala, for example, but I would want to be able to share variables between Scala chunks.

Interestingly, one of the reasons I was looking at quarto was because I thought it might be lighter in CI/CD pipelines using containers.  For example, I have a user guide written with bookdown, and I'm fiddling with a deployment to GitLab pages.  I've got a custom Docker image with R and bookdown installed, and it's currently about 600MB.  In the specific situation where I didn't need to actually evaluate R code, I wondered if I could make an image that had just Quarto and TinyTex that was smaller.  It's not a big deal--600MB is far from the biggest image I've ever see--but I was curious nonetheless.

**Reply by mgiugliano** [reply]
It would be truly fantastic to have multiple kernels per document: each cell is already bound to a single language.

Any possible workaround? I would love to use Python3 and Julia for writing a book or a paper, demonstrating computer simulations (involving heavy use of Python libraries) and otherwise showing Julia code for more hands-on numerical solution of equations.

Also incorporating another simultaneous kernel (say 'bash') would allow to demonstrate scripting in a reproducible way.

**Reply by pbowyer** [reply]
I really like this idea @mgiugliano. When documenting how to do something in multiple languages I always think of https://docusaurus.io/docs/markdown-features/code-blocks#multi-language-support-code-blocks as the ideal kind of output for the web version - especially how it remembers your language choice and changes all blocks. Taking this presentation and extending it to executing the code and showing the output would be phenomenal. Though if all are meant to produce the same output you might want to pick a single output to include in the document and suppress the others, to save space/files/weight.

**Reply by cmhh** [reply]
@pbowyer I love the idea of having the different language examples in tabs.  Give that julyter is the underlying engine, I suspect the feedback is going to be that it's technically challenging to mix languages in a single document.  But it would be an amazing feature for sure.

**Reply by mgiugliano** [reply]
@pbowyer @cmhh I must have initially misunderstood Quarto: I thought each "cell" was a stand-alone notebook. Instead I realise that a Quarto document is basically a Jupyter notebook, so each cell share the workspace/memory/variables/functions etc.

I see the advantages, but I wonder whether one could enforce the "one-cell --> one notebook" limit case.

**Reply by eitsupi** [reply]
> It would be truly fantastic to have multiple kernels per document: each cell is already bound to a single language.
> 
> Any possible workaround? I would love to use Python3 and Julia for writing a book or a paper, demonstrating computer simulations (involving heavy use of Python libraries) and otherwise showing Julia code for more hands-on numerical solution of equations.

As @cderv points out, just use the knitr engine instead of the jupyter engine.

In addition to R and knitr, you will need the reticulate package to run Python and the JuliaCall package to run Julia.

![image](https://user-images.githubusercontent.com/50911393/182830123-2177749d-6b66-46ea-a8c9-2892cd535b9d.png)

````qmd
---
title: multi-lang sample
engine: knitr
---

:::{.panel-tabset}

## python

```{python}
x = "this is python code."
```

## julia

```{julia}
x = "this is julia code."
```

:::

:::{.panel-tabset}

## python

```{python}
print(x)
```

## julia

```{julia}
println(x)
```

:::
````

**Reply by mgiugliano** [reply]
Thank you @eitsupi ! I will study that.

It is already sad (for me) that Quarto comes bundled with its own Pandoc. If I have to install R to use knitr it, to use Quarto, then it might be a no-go for me. I might be too sensitive to minimalism though :-)
